tests/dialects/test_memref.py

```diff
@@ -114,3 +118,66 @@ def test_memref_dealloca():
     dealloc0 = Dealloca.get(alloc0)
 
     assert type(dealloc0.memref) is OpResult
+
+
+def test_memref_dim():
+    idx = arith.Constant.from_int_and_width(1, IndexType())
+    alloc0 = Alloc.get(i32, 64, [3, 1, 2])
+    dim_1 = memref.Dim.from_source_and_index(alloc0, idx)
+
+    assert dim_1.source is alloc0.memref
+    assert dim_1.index is idx.result
+    assert isinstance(dim_1.result.typ, IndexType)
+
+
+def test_memref_rank():
+    alloc0 = Alloc.get(i32, 64, [3, 1, 2])
+    dim_1 = memref.Rank.from_memref(alloc0)
+
+    assert dim_1.source is alloc0.memref
+    assert isinstance(dim_1.rank.typ, IndexType)
+
+
+def test_memref_matmul():
+    memref_f64_rank2 = memref.MemRefType.from_element_type_and_shape(
+        builtin.f64, [-1, -1])
+
+    module = builtin.ModuleOp.from_region_or_ops([
+        func.FuncOp.from_callable(
+            'matmul',
+            [memref_f64_rank2, memref_f64_rank2],
+            [memref_f64_rank2],
+            lambda a, b: [
+                lit0 := arith.Constant.from_int_and_width(0, builtin.IndexType()),
+                lit1 := arith.Constant.from_int_and_width(1, builtin.IndexType()),
+                dim_a0 := memref.Dim.from_source_and_index(a, lit0),
+                dim_a1 := memref.Dim.from_source_and_index(a, lit1),
+                dim_b0 := memref.Dim.from_source_and_index(b, lit0),
+                dim_b1 := memref.Dim.from_source_and_index(b, lit1),
+                out := memref.Alloca.get(builtin.f64, 0, [-1, -1], [dim_a0, dim_b1]),
+                # TODO: assert dim_a0 == dim_b1
+                lit0_f := arith.Constant.from_float_and_width(0.0, builtin.f64),
+                func.Return.get(out)
+            ]
+        )
+    ])  # yapf: disable
```

Comment:
why disable yapf?

Reply:
Because it makes it unreadable otherwise :/